src/diffusers/callbacks.py

```diff
@@ -0,0 +1,116 @@
+from typing import List
+
+from .configuration_utils import ConfigMixin, register_to_config
+from .utils import CONFIG_NAME
+
+
+class PipelineCallback(ConfigMixin):
+    config_name = CONFIG_NAME
+
+    @register_to_config
+    def __init__(self, cutoff_step_ratio=1.0, cutoff_step_index=None):
+        super().__init__()
+
+        if (cutoff_step_ratio is None and cutoff_step_index is None) or (
+            cutoff_step_ratio is not None and cutoff_step_index is not None
+        ):
+            raise ValueError("Either cutoff_step_ratio or cutoff_step_index should be provided, not both or none.")
+
+        if cutoff_step_ratio is not None and (
+            not isinstance(cutoff_step_ratio, float) or not (0.0 <= cutoff_step_ratio <= 1.0)
+        ):
+            raise ValueError("cutoff_step_ratio must be a float between 0.0 and 1.0.")
+
+    @property
+    def tensor_inputs(self) -> List:
+        raise NotImplementedError(f"You need to set the attribute `tensor_inputs` for {self.__class__}")
+
+    def callback_fn(self, pipeline, step_index, timesteps, callback_kwargs):
+        raise NotImplementedError(f"You need to implement the method `callback_fn` for {self.__class__}")
+
+    def __call__(self, pipeline, step_index, timestep, callback_kwargs):
```

Comment:
Just from my past experience, I found it helpful to have callback/hook signature be more flexible by adding `*args, **kwargs`. Otherwise, when at any point in the future, we find that we need an additional argument for a specific use case and add it, all existing callbacks would break unless updated.

Of course, since the callback system is already in place for some time, you may know better than me that this requirement will never change, I just wanted to highlight this possibility.

Reply:
I also do this but I read a while ago that the general approach on `diffusers` about `*args` and `**kwargs` was not to use them. @yiyixuxu WDYT?

Reply:
I think in this case, the signature for callback is highly unlikely to change though
This is how the callback is invoked inside the pipeline https://github.com/huggingface/diffusers/blob/75aab34675bf22a8ad12066cf0178c7c7c79bdd0/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L1004

the `callback_kwargs` is a dict containing variables that the user defines, and is different for each callback